Possible incorrect behavior in checkAndPut

From https://stackoverflow.com/questions/44174094/checkandput-always-succeeds

There are only tests for equality comparators in checkAnd*.
Add tests for "<" or ">" (non commutative comparators) and verify that their behavior is as defined in the documentation:
"GREATER operator means expected value > existing <=> add the put."

@sduskis There are still counter-examples wherein `CompareOp` doesn't work correctly.  Here is a test against master which clearly should pass but it fails,

```
@Test
  public void testCompareOpGreater() throws IOException {
    byte[] rowKey = dataHelper.randomData("rowKey-");
    byte[] qualToCheck = dataHelper.randomData("toCheck-");
    boolean success;

    Table table = getConnection().getTable(sharedTestEnv.getDefaultTableName());

    table.put(new Put(rowKey).addColumn(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck,
            Bytes.toBytes(0l)));

    Assert.assertEquals(0, Bytes.toLong(table.get(new Get(rowKey)).getValue(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck)));

    success = table.checkAndPut(rowKey, SharedTestEnvRule.COLUMN_FAMILY, qualToCheck,
            CompareOp.GREATER, Bytes.toBytes(150l), new Put(rowKey).addColumn(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck, Bytes.toBytes(200l)));
    Assert.assertTrue("150 > 0 should succeed", success);

    Assert.assertEquals(200, Bytes.toLong(table.get(new Get(rowKey)).getValue(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck)));

    success = table.checkAndPut(rowKey, SharedTestEnvRule.COLUMN_FAMILY, qualToCheck,
            CompareOp.GREATER, Bytes.toBytes(100l), new Put(rowKey).addColumn(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck, Bytes.toBytes(300l)));
    Assert.assertFalse("100 > 200 should fail", success);
  }
```

Add the above to `TestCheckAndMutate` and run against master (using emulator),

```
mvn -Dgoogle.bigtable.project.id=project_id -Dgoogle.bigtable.instance.id=instance_id -Dgoogle.bigtable.connection.impl=com.google.cloud.bigtable.hbase1_x.BigtableConnection  -Dtest=TestCheckAndMutate#testCompareOpGreater clean test

Running com.google.cloud.bigtable.hbase.TestCheckAndMutate
2017-06-30 19:44:41 DEBUG Recycler:71 - -Dio.netty.recycler.maxCapacityPerThread: disabled
2017-06-30 19:44:41 DEBUG Recycler:71 - -Dio.netty.recycler.maxSharedCapacityFactor: disabled
2017-06-30 19:44:41 DEBUG Recycler:71 - -Dio.netty.recycler.linkCapacity: disabled
2017-06-30 19:44:41 DEBUG Recycler:71 - -Dio.netty.recycler.ratio: disabled
2017-06-30 19:44:41 INFO  TestCheckAndMutate:45 - Running: testCompareOpGreater(com.google.cloud.bigtable.hbase.TestCheckAndMutate)
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.593 sec <<< FAILURE! - in com.google.cloud.bigtable.hbase.TestCheckAndMutate
testCompareOpGreater(com.google.cloud.bigtable.hbase.TestCheckAndMutate)  Time elapsed: 0.122 sec  <<< FAILURE!
java.lang.AssertionError: 100 > 200 should fail
	at com.google.cloud.bigtable.hbase.TestCheckAndMutate.testCompareOpGreater(TestCheckAndMutate.java:313)


Results :

Failed tests:
  TestCheckAndMutate.testCompareOpGreater:313 100 > 200 should fail
```

--- a/bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestCheckAndMutate.java
+++ b/bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestCheckAndMutate.java
@@ -352,4 +352,27 @@ public class TestCheckAndMutate extends AbstractTest {
       CompareOp.NOT_EQUAL, Bytes.toBytes(4000l), someRandomPut);
     Assert.assertTrue("4000 != 2000 should succeed", success);
   }
+
+  @Test
+  public void testCompareOpsVersions() throws IOException {
+    byte[] rowKey = dataHelper.randomData("rowKey-");
+    byte[] qualToCheck = dataHelper.randomData("toCheck-");
+    byte[] otherQual = dataHelper.randomData("other-");
+    boolean success;
+
+    Table table = getConnection().getTable(sharedTestEnv.getDefaultTableName());
+    Put someRandomPut =
+        new Put(rowKey).addColumn(SharedTestEnvRule.COLUMN_FAMILY, otherQual, Bytes.toBytes(1l));
+
+    table.put(new Put(rowKey, System.currentTimeMillis() - 10000)
+        .addColumn(SharedTestEnvRule.COLUMN_FAMILY, qualToCheck, Bytes.toBytes(2000l)));
+
+    table.put(new Put(rowKey, System.currentTimeMillis()).addColumn(SharedTestEnvRule.COLUMN_FAMILY,
+      qualToCheck, Bytes.toBytes(4000l)));
+
+    success = table.checkAndPut(rowKey, SharedTestEnvRule.COLUMN_FAMILY, qualToCheck,
+      CompareOp.GREATER, Bytes.toBytes(3000l), someRandomPut);
+    Assert.assertFalse("3000 > 4000 should fail", success);
+
+  }
 }

--- a/bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java
+++ b/bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java
@@ -96,6 +96,8 @@ public class ScanAdapter implements ReadOperationAdapter<Scan> {
       chainBuilder.addFilters(createTimeRangeFilter(scan.getTimeRange()));
     }
 
+    chainBuilder.addFilters(createColumnLimitFilter(scan.getMaxVersions()));
+
     if (scan.getFilter() != null) {
       Optional<RowFilter> userFilter = createUserFilter(scan, hooks);
       if (userFilter.isPresent()) {
@@ -103,7 +105,6 @@ public class ScanAdapter implements ReadOperationAdapter<Scan> {
       }
     }
 
-    chainBuilder.addFilters(createColumnLimitFilter(scan.getMaxVersions()));
 
     if (chainBuilder.getFiltersCount() == 1) {
       return chainBuilder.getFilters(0);

